### ISSUE TITLE
Widgets deleted during execution of PyDM can leave dangling signal connections

### ISSUE THREAD

**Issue by ZLLentz** [slaclab/pydm#798]
**Describe the bug**

The following sequence leaves a dangling signal connection, which can continue to emit long after the widget is deallocated:
1. Create two widgets with the same `PyDMChannel` address in the same application
2. Delete one of these widgets using `QObject.deleteLater()`

This is because deleting a widget causes the following sequence:
1. emit the `destroyed` signal, which we set up here: https://github.com/slaclab/pydm/blob/32bd344243bc9eedd860c6eebd130b3e531d7b8b/pydm/widgets/base.py#L602
2. call `widget_destroyed`: https://github.com/slaclab/pydm/blob/32bd344243bc9eedd860c6eebd130b3e531d7b8b/pydm/widgets/base.py#L82
3. call `disconnect` with `destroying=True`, which eventually makes its way to `remove_listener` with `destroying=True` at https://github.com/slaclab/pydm/blob/32bd344243bc9eedd860c6eebd130b3e531d7b8b/pydm/data_plugins/plugin.py#L80
4. The apparent affect of `destroying=True` is to skip the part where we disconnect the signals from the slots.

**Expected behavior**

I would expect my widget's signals and slots to be disconnected when the widget is destroyed.

I'm making this issue because I'm really confused- an argument called `destroying` results in us NOT cleaning up the object. Is there any situation where it is preferable for the widget's signals to not be cleaned up when we destroy it, or even any time when we're disconnecting a `PyDMChannel` where we want the signals and slots to stay connected?

Is there any documentation on the intent of this `destroying` kwarg?

**Steps to Reproduce**

Smallest reproduction is probably:
1. Create two `PyDMChannel` objects with the same address and connect them, link the value slot to a test function
2. Call `.disconnect(destroying=True)` on the one connected to the test function
3. Update the value of the channel
4. Note that your test function is called even though the associated channel has been disconnected

**Possible Solution**
 
- Make `widget_destroyed` use `destroying=False`?
- Deprecate and ignore the `destroying` kwarg, always disconnecting the signals?
- Add an option for a `PyDMWidget` or a `PyDMChannel` to disconnect the signals on cleanup?

**My Platform**

Red Hat 7
Python 3.9.6
Pydm 1.13.0
qt 5.12.9
pyqt 5.12.3

**Additional context**

I ran into this while making custom widgets that are subclasses of `PyDMWidget`. My only apparent recourse is to wrap some of my code in extra try/except blocks and accept that dynamically deleting widgets has some minor performance consequences.

**Comment by jbellister-slac** [comment]
Hi Zach,

Apologies for the delay on getting to this one. I didn't see any documentation for the `destroying` kwarg either, so I tried just ignoring it and ended up with a lot of this getting printed out when deleting a PyDMWidget:

`QObject::disconnect: Unexpected null parameter`

The reason being that Qt will automatically clean up the signals and slots of a QObject for you when destroying it (https://doc.qt.io/qt-5/qobject.html#dtor.QObject) so by trying to do it again we end up with a bunch of errors in the log. So the intent of the parameter is that if we are coming from a flow in which the PyDMWidget is being destroyed, there is no need to do any disconnecting ourselves because it is automatically handled by Qt.

I confirmed this with a minimal example where clicking the button and deleting the label does clean everything up without any errors

    def __init__(self, parent=None, args=None, macros=None):
        super().__init__(parent=parent, args=args, macros=None)
        self.my_layout = QVBoxLayout(self)
        self.button = PyDMPushButton()
        self.button.clicked.connect(self.delete_label)

        self.my_layout.addWidget(self.button)

        self.label = PyDMLabel(init_channel='ca://MTEST:MeanValue')
        self.label2 = PyDMLabel(init_channel='ca://MTEST:MeanValue')
        self.my_layout.addWidget(self.label)
        self.my_layout.addWidget(self.label2)

    def delete_label(self):
        self.label.deleteLater()`

So I took a look at your code and was able to reproduce the issue by using functools.partial to create a slot, like you do when creating the PyDMChannel in TyphosAlarm. Apparently partial is causing a problem with Qt where it won't automatically disconnect that signal/slot upon destroying the widget.

I can take a look at handling this, but based on my testing that appears to be the specific issue we're running into here.
